src/features/devices/models/device-additions.ts

```diff
@@ -4,8 +4,139 @@ import type {
 	AndNode,
 	BooleanTypeNodes,
 	EqualsNode,
+	OrNode,
 } from '@balena/abstract-sql-compiler';
 
+// The Cloud API has the notion of active/inactive devices via a field
+// on the device resource. Computing the overall status below depends
+// on whether the device 'is active'.
+//
+// We must somehow expose the overall status/progress attributes on the
+// model and we could contrive a way for the cloud to inject the extra
+// cases, but that would quickly become overly complex because the order
+// of cases matters. We'd never be able to cleanly make the core API
+// agnostic to the cloud.
+//
+// Therefore, we opted for using the 'is active' field when it's part of
+// the model, otherwise we use a dummy `false` value everywhere that field
+// is referenced. This approach can also be used in core API
+// translations in a consistent way.
+export const isInactiveFn = (isActiveNotExist: boolean): BooleanTypeNodes =>
+	isActiveNotExist
+		? ['Boolean', false]
+		: ['Not', ['ReferencedField', 'device', 'is active']];
+
+export const isOverallOffline: AndNode = [
+	'And',
+	['Not', ['ReferencedField', 'device', 'is online']],
+	[
+		'In',
+		['ReferencedField', 'device', 'api heartbeat state'],
+		['EmbeddedText', 'offline'],
+		['EmbeddedText', 'unknown'],
+	],
+];
+
+// VPN not connected and supervisor didn't yet reach the API,
+// so it's still provisioning.
+export const isPreProvisioning: AndNode = [
+	'And',
+	['Not', ['ReferencedField', 'device', 'is online']],
+	['NotExists', ['ReferencedField', 'device', 'last connectivity event']],
+	[
+		'Equals',
+		['ReferencedField', 'device', 'api heartbeat state'],
+		['EmbeddedText', 'unknown'],
+	],
+];
+
+export const isPostProvisioning: EqualsNode = [
+	'Equals',
+	['ReferencedField', 'device', 'provisioning state'],
+	['EmbeddedText', 'Post-Provisioning'],
+];
+// This check does not double check heartbeat state as it is already checked from isOverallOffline which runs before
+export const hasPartialConnectivity: OrNode = [
```

Comment:
How about v so that vpn disabled&off & heartbeat timeout gives reduced-functionality?
```
-- In an ideal world:
partial = CASE
	WHEN vpn disabled
	THEN heartbeat = timeout
	ELSE (vpn on & heartbeat != online) || (vpn off & heartbeat = online,timeout)

=>

partial = CASE
	WHEN heartbeat = timeout THEN True
	ELSE (vpn enabled && vpn on & heartbeat IN offline)
		|| (vpn off & heartbeat = online && vpn enabled)

approximately =>

partial = CASE
	WHEN heartbeat = timeout THEN True
	ELSE (vpn on & heartbeat IN offline)
		|| (vpn off & heartbeat = online && vpn enabled)

=>

partial =
	|| heartbeat = timeout
	|| (vpn on & heartbeat IN offline/unknown)
	|| (vpn off & heartbeat = online && vpn enabled)
```